src/core/coap/coap_message.cpp

```diff
@@ -358,11 +358,30 @@ otError Message::ParseHeader(void)
     VerifyOrExit(GetHelpData().mNextOptionOffset > 0, error = OT_ERROR_PARSE);
     GetHelpData().mHeaderLength = GetHelpData().mNextOptionOffset - GetHelpData().mHeaderOffset;
     MoveOffset(GetHelpData().mHeaderLength);
+    GetHelpData().mIsMessageIdValid = true;
 
 exit:
     return error;
 }
 
+uint16_t Message::GetMessageId(void) const
+{
+    const HelpData &helpData = GetHelpData();
+
+    // Invalid usage to get before set
+    assert(helpData.mIsMessageIdValid);
+
+    return HostSwap16(helpData.mHeader.mMessageId);
+}
+
+void Message::SetMessageId(uint16_t aMessageId)
```

Comment:
In the file `border_agent.cpp`, there are two calls to `SetMessageId` I think might have issues.

 - `message->SetMessageId(aSeparate ? 0 : aRequest.GetMessageId());` in `SendErrorMessage`
 - `aMessage.SetMessageId(mSeparate ? 0 : mMessageId);` in `ToHeader`

These two `SetMessageId` might set the message ID to 0, which used to means `using the next message ID`. 
Please check if these two calls is still valid after the semantic change of message ID being zero.

Reply:
They are correct. The message id is set `0` when this is a separate response. For piggy-backed response, the message id should be identical to that of the original request.

Reply:
I got your idea.